### PR TITLE
[FLINK-28778][SQL/API] Bulk fetch of table and column statistics for given partitions

### PR DESCRIPTION
**Pull request by luoyuxia**
## What is the purpose of the change
A continue work for #20422 .

Currently the statistics information about tables can only be fetched from the catalog by each given partition iteratively. Since getting statistics information from catalogs is a very heavy operation, in order to improve the query performance, we’d better provide functionality to fetch the statistics information of a table for all given partitions in one shot.

Based on the manual performance test, for 2000 partitions, the cost will be improved from 10s to 2s. The improvement result is 500%.

## Brief change log
  -   Implementations for new added methods will be done in all classes that implement the Catalog interface.
  -  For the concrete HiveCatalog, HIVE_DEFAULT_PARTITION will be taken care of while fetching the column statistics. - - All currently supported types for partition column will be taken into consideration.
  - The logic of getting TableStates in FlinkRecomputeStatisticsProgram will be optimized from calling the catalog iteratively to bulk fetch.
  - Logic of TableStates conversion will be consolidated into the CatalogTableStatisticsConverter class to make the domain design a little bit cleaner.

## Verifying this change
Added test in `HiveCatalogHiveMetadataTest` and `GenericInMemoryCatalogTest`.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (not applicable / docs / JavaDocs / **not documented**)

### CODE REVIEW

**Comment by godfreyhe** [comment]
Please refer https://github.com/apache/flink/pull/20422 to see the previous comments

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/catalog/stats/CatalogTableStatistics.java [hunk 1]
```diff
@@ -27,8 +27,11 @@
 /** Statistics for a non-partitioned table or a partition of a partitioned table. */
 @PublicEvolving
 public class CatalogTableStatistics {
+
     public static final CatalogTableStatistics UNKNOWN = new CatalogTableStatistics(-1, -1, -1, -1);
 
+    public static final CatalogTableStatistics EMPTY = new CatalogTableStatistics(0, 0, 0, 0);
```
Comment (godfreyhe) [inline flink-table/flink-table-common/src/main/java/org/apache/flink/table/catalog/stats/CatalogTableStatistics.java:33]: useless line, remove

File: flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/catalog/hive/util/HiveStatsUtil.java [hunk 2]
```diff
@@ -90,6 +102,257 @@ public static Map<String, CatalogColumnStatisticsDataBase> createCatalogColumnSt
...
+
+    /**
+     * Get statistics for specific partition column.
+     *
+     * @param logicalType the specific partition column's logical type
+     * @param partitionValue the partition value for the specific partition column
+     * @param partitionColIndex the index of the specific partition
+     * @param defaultPartitionName the default partition name for null value
+     */
+    private static CatalogColumnStatisticsDataBase getColumnStatistics(
```
Comment (godfreyhe) [inline flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/catalog/hive/util/HiveStatsUtil.java:171]: how about `getPartitionColumnStats` ?

File: flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/catalog/hive/util/HiveStatsUtil.java [hunk 3]
```diff
@@ -90,6 +102,257 @@ public static Map<String, CatalogColumnStatisticsDataBase> createCatalogColumnSt
...
+                                getNullCount(
+                                        client, hiveTable, partitionColIndex, defaultPartitionName);
+                    } else {
+                        min = new Date(((LocalDate) partitionValue).toEpochDay());
+                        max = min;
+                    }
+                    return new CatalogColumnStatisticsDataDate(min, max, 1L, nullCount);
+                }
+            default:
+                return null;
```
Comment (godfreyhe) [inline flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/catalog/hive/util/HiveStatsUtil.java:267]: TIMESTAMP is unsupported ?

File: flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/catalog/hive/util/HiveStatsUtil.java [hunk 4]
```diff
@@ -90,6 +102,257 @@ public static Map<String, CatalogColumnStatisticsDataBase> createCatalogColumnSt
...
+
+    /**
+     * Get the null count for the {@param partitionColIndex} partition column in table {@param
+     * hiveTable}.
+     *
+     * <p>To get the null count, it will first list all the partitions whose {@param
+     * partitionColIndex} partition column is null, and merge the partition's statistic to get the
+     * total rows, which is exactly null count for the {@param partitionColIndex} partition column.
+     */
+    private static Long getNullCount(
```
Comment (godfreyhe) [inline flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/catalog/hive/util/HiveStatsUtil.java:279]: getPartitionColumnNullCount

File: flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/catalog/hive/util/HiveStatsUtil.java [hunk 5]
```diff
@@ -90,6 +102,257 @@ public static Map<String, CatalogColumnStatisticsDataBase> createCatalogColumnSt
         return colStats;
     }
 
+    public static Map<String, CatalogColumnStatisticsDataBase> createCatalogPartitionColumnStats(
```
Comment (godfreyhe) [inline flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/catalog/hive/util/HiveStatsUtil.java:105]: getCatalogPartitionColumnStats ?

**Review by godfreyhe — COMMENTED** [review]
Thanks for the contribution, I left some comments. Could you add some test to verify the default implementation of new methods introduced in Catalog, and update the pr description

File: flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/catalog/hive/util/HiveStatsUtil.java (hunk 3 above)
Comment (luoyuxia) [reply flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/catalog/hive/util/HiveStatsUtil.java:278]: Miss it. `TIMESTAMP` should be supported.

**Comment by luoyuxia** [comment]
@godfreyhe Thanks for your review. I have addressed all your comments.

**Review by godfreyhe — APPROVED** [review]
Thanks for the update, LGTM overall. It's very nice that we can add some cases for Timestamp

**Comment by godfreyhe** [comment]
The failure test is due to https://issues.apache.org/jira/browse/FLINK-28844

**Pull request closed without merging**